### ISSUE TITLE
False negative for E1012 Ref not found

### ISSUE THREAD

**Issue by adamchainz** [aws-cloudformation/cfn-python-lint#1955]
*cfn-lint version: (`cfn-lint --version`)* 0.48.0

*Description of issue.*

Example template:

```yaml
AWSTemplateFormatVersion: 2010-09-09

Resources:

  A:
    Type: AWS::Route53::HostedZone
    Properties:
      Name: example.com.

  Record:
    Type: AWS::Route53::RecordSet
    Properties:
      Name: example.com.
      Type: CNAME
      HostedZoneId: !Ref AB
      ResourceRecords:
      - example.org
```

Linting this passes even though the `!Ref AB` fails on cloudformation

Commeting out resource "A" makse it fail - it looks to me like any `!Ref` that matches a prefix will pass. This may be due to use of `re.match` in `RegexDict` without a terminating `$`.

**Comment by kddejong** [comment]
Good catch. I'm on it.

**Comment by PatMyron** [comment]
> it looks to me like any `!Ref` that matches a prefix will pass

@kddejong is this from modules support? Should continue matching prefixes for modules while only allowing exact matches for resources

**Comment by kddejong** [comment]
yea, I just got a PR in to do exactly that.

**Comment by kddejong** [comment]
Looks like the RegexDict fixes we did for modules opened up this issue.  Switching that Regex dict to handle only doing that comparison when the type is a Module.
